pkg/network/ebpf/c/prebuilt/http.c

```diff
@@ -63,14 +63,37 @@ int socket__http_filter(struct __sk_buff* skb) {
     return 0;
 }
 
-// This kprobe is used to send batch completion notification to userspace
-// because perf events can't be sent from socket filter programs
+SEC("kprobe/tcp_sendmsg")
+int kprobe__tcp_sendmsg(struct pt_regs* ctx) {
+    // map connection tuple during SSL_do_handshake(ctx)
+    init_ssl_sock_from_do_handshake((struct sock*)PT_REGS_PARM1(ctx));
+    return 0;
+}
+
 SEC("kretprobe/tcp_sendmsg")
 int kretprobe__tcp_sendmsg(struct pt_regs* ctx) {
+    // send batch completion notification to userspace
+    // because perf events can't be sent from socket filter programs
     http_notify_batch(ctx);
     return 0;
 }
 
+
+SEC("uprobe/SSL_do_handshake")
+int uprobe__SSL_do_handshake(struct pt_regs* ctx) {
+    u64 pid_tgid = bpf_get_current_pid_tgid();
+    void *ssl_ctx = (void *)PT_REGS_PARM1(ctx);
+    bpf_map_update_elem(&ssl_ctx_by_pid_tgid, &pid_tgid, &ssl_ctx, BPF_ANY);
```

Comment:
Is `SSL_do_handshake` single-threaded? What guarantees do we have that the `tcp_sendmsg` call will be from this context, other than being the same PID and Thread Group ID?

Reply:
As discuss offline, (on haproxy server side) `SSL_do_handshake` will use a memory BIO and hook is own method/function helper to write to the destination socket (only haproxy have the context between *ssl_ctx and fd(socket)). As openssl is a C library we can say the caller thread (SSL_do_handshake) will match the syscall (send) thread, and then tcp_sendmsg.      The only issue (not covered by this PR) it's when openssl async mode is enabled (via haproxy.global.ssl-async-mode) as SSL_do_handshake() will return early.